### PR TITLE
Adding a way to override feature flags (for commissioner discovery) for Python chip-device-ctrl tool

### PR DESCRIPTION
**Pull request by sharadb-amazon**
#### Problem
The flags defined in src/platform/include/CHIPDeviceConfig.h could be overriden by various example apps, but not for the python controller. And we needed a way to set the commissioner discovery flags to true for the python controller.

#### Change overview
* Creates a way to override feature flags for the python controller.
* Sets CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY and CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE to 1

#### Testing
Tested against the minimal-mdns client and verified that the py chip-device-ctrl tool advertises itself as a commissioner after this config change.

### CODE REVIEW

**Review by mspang — DISMISSED** [review]
Adding a new .gn file is for establishing an entirely new project. We use this from examples/ because those are examples of making a new project.

The python controller is part of the main project, and shouldn't have it's own build system.

**Comment by andy31415** [comment]
Why do we need a compile-time configuration for this in python?

I would generally expect python to have 'enable everything' and then at runtime to be able  to select if something is used or not.

It seems to me like user defined commisioning compile-time enabling or not has more deep sideffects in functionality...maybe it should not.

@sharadb-amazon - could you explain why we need a compile-time flag for python? What do these compile flags enable and why can't we have python build with 'enabled' and change functionality via methods (some true/false enabling or other arguments)?

**Comment by sharadb-amazon** [comment]
> Why do we need a compile-time configuration for this in python?
> 
> I would generally expect python to have 'enable everything' and then at runtime to be able to select if something is used or not.
> 
> It seems to me like user defined commisioning compile-time enabling or not has more deep sideffects in functionality...maybe it should not.
> 
> @sharadb-amazon - could you explain why we need a compile-time flag for python? What do these compile flags enable and why can't we have python build with 'enabled' and change functionality via methods (some true/false enabling or other arguments)?

CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY and CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE are pre-existing compile time flags that are used in the CHIPDeviceController.cpp (and elsewhere) to guard code that implements user directed commissioning and commissioner advertisement in the SDK. These need to be enabled *only* for the python controller (disabled by default for example apps) so that it responds to commissioner discovery and UDC requests.

**Comment by mspang** [comment]
> > Why do we need a compile-time configuration for this in python?
> > I would generally expect python to have 'enable everything' and then at runtime to be able to select if something is used or not.
> > It seems to me like user defined commisioning compile-time enabling or not has more deep sideffects in functionality...maybe it should not.
> > @sharadb-amazon - could you explain why we need a compile-time flag for python? What do these compile flags enable and why can't we have python build with 'enabled' and change functionality via methods (some true/false enabling or other arguments)?
> 
> CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY and CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE are pre-existing compile time flags that are used in the CHIPDeviceController.cpp (and elsewhere) to guard code that implements user directed commissioning and commissioner advertisement in the SDK. These need to be enabled _only_ for the python controller (disabled by default for example apps) so that it responds to commissioner discovery and UDC requests. Note, these are NOT new or python specific flags.

Those defines are part of the implementation of this feature. Even if it's already merged in the tree, it's still subject to scrutiny later when we go to adjust configurations. 

I don't agree that the Linux samples should be forced to statically enable or disable these features. It isn't a scalable approach and will cause a huge amount of combinatorial pain as we try to test different cases of optional features being enabled or disabled.

File: scripts/build_python.sh [hunk 1]
```diff
@@ -102,7 +102,7 @@ source "$CHIP_ROOT/scripts/activate.sh"
 # Generates ninja files
 [[ -n "$chip_mdns" ]] && chip_mdns_arg="chip_mdns=\"$chip_mdns\"" || chip_mdns_arg=""
 
-gn --root="$CHIP_ROOT" gen "$OUTPUT_ROOT" --args="chip_detail_logging=$chip_detail_logging enable_pylib=$enable_pybindings enable_rtti=$enable_pybindings chip_use_clusters_for_ip_commissioning=$clusters $chip_mdns_arg"
+gn --root="$CHIP_ROOT" gen "$OUTPUT_ROOT" --args="chip_detail_logging=$chip_detail_logging enable_pylib=$enable_pybindings enable_rtti=$enable_pybindings chip_use_clusters_for_ip_commissioning=$clusters chip_system_project_config_include_dirs=\"/config/python\" $chip_mdns_arg"
```
Comment (mspang) [inline scripts/build_python.sh:105]: Doesn't this work already?

```
chip_project_config_include_dirs = [ "//config/python" ]
```

**Review by chrisdecenzo — APPROVED** [review]

**Comment by mspang** [comment]
I can help you make a targeted compile time argument for this to unblock the test event, but I would advocate strongly for the position that **optional features be able to be disabled without recompiling the code** and that Linux should by default have a maximalist compile-time configuration and enable or disable features as appropriate at runtime. 

> #### Problem
> The flags defined in src/platform/include/CHIPDeviceConfig.h could be overriden by various example apps, but not for the python controller. And we needed a way to set the commissioner discovery flags to true for the python controller.

While they 'can' be overridden by sample apps, we were not using this capability on Linux until recently. I actually don't agree with the changes that were made here in the lighting, tv, and tv-casting apps - these changes only apply when building from the subdirectory and we also support building them at the top level.

> 
> #### Change overview
> * Creates a way to override feature flags for the python controller.
> * Sets CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY and CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE to 1
> 
> #### Testing
> Tested against the minimal-mdns client and verified that the py chip-device-ctrl tool advertises itself as a commissioner after this config change.

**Review by mspang — APPROVED** [review]

**Comment by sharadb-amazon** [comment]
> **optional features be able to be disabled without recompiling the code** and that Linux should by default have a maximalist compile-time configuration and enable or disable features as appropriate at runtime.

Opened an issue to track this work: https://github.com/project-chip/connectedhomeip/issues/9889

**Review by hawk248 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by franck-apple** [comment]
This seems to be breaking BLE provisioning in the TE6 branch. Need to remove it from the TE6 branch.

**Comment by andy31415** [comment]
Revert done.

**Pull request merged**